src/conversions.jl

```diff
@@ -0,0 +1,302 @@
+
+    ind1 = Int[]
+    ind2 = Int[]
+    ind3 = Int[]
+
+    for x = 1:resolution, y = 1:resolution, z = 1:resolution
+        push!(ind1, x)
+        push!(ind2, y)
+        push!(ind3, z)
+    end
+    ind1 = reshape(ind1, 1, :)
+    ind2 = reshape(ind2, 1, :)
+    ind3 = reshape(ind3, 1, :)
+
+    grid_points = cat(ind1, ind2, ind3, dims = 1)
+    grid_points = ((grid_points .+ 0.5) ./ resolution)
+
+    nn_idx = cat(
+        [
+            CartesianIndex.(
+                reduce(vcat, knn(KDTree(cloud[:, :, i]), grid_points, 1)[1]),
+                i,
+            ) for i = 1:_B
+        ]...,
+        dims = 2,
+    )
+
+    dists_vec = (grid_points .- cloud[:, nn_idx])
+    dists = sum((dists_vec .^ 2), dims = 1)
+    dists = reshape(dists, resolution, resolution, resolution, _B)
+    voxels = typeof(similar(pcloud.points,1,1,1,1))(dists .<= (0.6 / (resolution * resolution)))
+    return voxels
+end
+
+function trimesh_to_voxel(m::TriMesh{T,R,S},res::Int=32)    where {T,R,S}
+    verts_list = get_verts_list(m)
+    faces_list = get_faces_list(m)
+
+    _B = length(verts_list)
+    voxels = S{T,4}(undef,res,res,res,_B)
+    # print(typeof(verts_list))
+    for (i,v,f) in zip(1:_B,verts_list, faces_list)
+        voxels[:,:,:,i] = _voxelize(cpu(v),f,res)
+    end
+    return voxels
+end
+
+function _voxelize(
+    v::Array{<:AbstractFloat,2},
+    f::Array{<:Integer,2},
+    resolution::Int = 32,
+)
+    verts_max = maximum(v)
+    verts_min = minimum(v)
+    verts = ((v .- verts_min) ./ (verts_max - verts_min))
+
+    points = verts
+    smallest_side = (1.0 / resolution)^2
+
+    v1 = verts[:, f[1, :]]
+    v2 = verts[:, f[2, :]]
+    v3 = verts[:, f[3, :]]
+
+    while true
+        side_1 = sum((v1 .- v2) .^ 2, dims = 1)
+        side_2 = sum((v2 .- v3) .^ 2, dims = 1)
+        side_3 = sum((v3 .- v1) .^ 2, dims = 1)
+        sides = cat(side_1, side_2, side_3, dims = 1)
+        sides = reshape(maximum(sides, dims = 1), :)
+
+        keep = sides .> smallest_side
+        if !(any(keep))
+            break
+        end
+
+        v1 = v1[:, keep]
+        v2 = v2[:, keep]
+        v3 = v3[:, keep]
+
+        v4 = (v1 + v3) / 2
+        v5 = (v1 + v2) / 2
+        v6 = (v2 + v3) / 2
+
+        points = cat(points, v4, v5, v6, dims = 2)
+        vertex_set = [v1, v2, v3, v4, v5, v6]
+        new_traingles = [
+            1 4 5
+            5 2 6
+            5 4 6
+            4 3 6
+        ]
+        new_verts = []
+        for i = 1:4
+            for j = 1:3
+                if i == 1
+                    push!(new_verts, vertex_set[new_traingles[i, j]])
+                else
+                    new_verts[j] = cat(
+                        new_verts[j],
+                        vertex_set[new_traingles[i, j]],
+                        dims = 2,
+                    )
+                end
+            end
+        end
+
+        v1, v2, v3 = new_verts
+    end
+
+    voxels = zeros(resolution, resolution, resolution)
```

Comment:
Might be an issue for cuarrays

Reply:
This function only expects array and later we convert to CuArray according to the original input. The problem with cuarray in this function was that logical indexing is not supported by it, and some lower function also uses other libraries like Meshing.jl which again expects Array.

Reply:
Hmm, is there advantage in having meshing.jl have cuda compatible kernels?

Reply:
Meshing.jl is fast even on cpu. For average size TriMesh (2000 verts) converting to voxels takes ~1ms.

Reply:
I am okay with having only cpu versions for now. These functions are generally used once for the lifetime of a project and are rarely a bottleneck. However, I believe opening an issue (or ask on slack) in Meshing.jl to ask if there are plans to include cuda kernels in the future would be worthwhile.